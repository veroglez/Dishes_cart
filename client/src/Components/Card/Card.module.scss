.card {
  flex: 1 0 calc(33% - 20px);
  border: 1px solid #eeeeee;
  margin-right: 20px;
  margin-bottom: 20px;

  .info {
    padding: 20px;

    .title {
      margin-top: 0;
    }

    .price {
      font-weight: 500;
    }
  }

  &:nth-child(3n) {
    margin-right: 0;
  }

  .image {
    width: 100%;
  }
}