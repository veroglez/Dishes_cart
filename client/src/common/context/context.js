import { node } from 'prop-types';
import React, { createContext, Component } from 'react';
import l10n from '../l10n';

const Context = createContext();
const { Provider, Consumer: ConsumerData } = Context;

class ProviderData extends Component {
  static propTypes = {
    children: node,
  };

  static defaultProps = {
    children: undefined,
  };

  state = {
    l10n,
    products: [],
    items: 0,
    total: 0,
  }

  onData = object => this.setState(object);

  render() {
    const {
      onData, props: { children },
    } = this;

    return (
      <Provider value={{ ...this.state, onData }}>
        { children }
      </Provider>
    );
  }
}

export { ConsumerData, ProviderData };