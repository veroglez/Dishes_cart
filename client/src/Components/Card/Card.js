import React, { PureComponent } from 'react';
import { number, string } from 'prop-types';
import { Price } from '..';
import { ConsumerData } from '../../common/context/context';
import style from './Card.module.scss';

class Card extends PureComponent {
  static propTypes = {
    image: string,
    index: number,
    title: string,
    price: number,
  };

  static defaultProps = {
    image: undefined,
    index: undefined,
    title: undefined,
    price: undefined,
  };

  addProduct = ({ products, total, onData }, index, title, image, price) => {
    let dish = {};
    const productInList = products.find(product => product.index === index);

    if (productInList) {
      productInList.quantity += 1;
    } else {
      dish = {
        index, title, image, price, quantity: 1,
      };
    }

    const totalPrice = total + price;
    onData({ products: productInList ? products : [...products, dish], total: totalPrice });
  };

  render() {
    const {
      addProduct,
      props: {
        image, index, title, price,
      },
    } = this;

    return (
      <ConsumerData>
        {({ l10n, ...context }) => (
          <div className={style.card}>
            <img className={style.image} src={image} alt={title} />
            <div className={style.info}>
              <p className={style.title}>{title}</p>
              <Price price={price} />
              <button
                type="button"
                onClick={() => addProduct(context, index, title, image, price)}
              >
                {l10n.ADD}
              </button>
            </div>
          </div>
        )}
      </ConsumerData>
    );
  }
}

export default Card;
