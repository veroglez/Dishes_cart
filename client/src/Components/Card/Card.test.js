import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Card from './Card';
import { ProviderData } from '../../common/context/context';

Enzyme.configure({ adapter: new Adapter() });

const DEFAULTS = {l10n: require('../../common/l10n')}; // eslint-disable-line
const IMAGE = '/image.png';
const PRICE = 23;
const TITLE = 'Lorem ipsum';

const providerData = shallow(<ProviderData />);
const card = shallow(<Card price={PRICE} />);
const { addProduct } = card.instance();
const { onData } = providerData.instance();
const total = 1000;
const products = [
  {
    index: 1, title: 'Lasagna', image: 'image.jpg', price: 595, quantity: 1,
  },
  {
    index: 2, title: 'Spaguetti', image: 'image.jpg', price: 234, quantity: 2,
  },
];


const HOC = inherit => (
  <ProviderData>
    <Card {...DEFAULTS} {...inherit} price={PRICE} />
  </ProviderData>
);


describe('<Card>', () => {
  it('render', () => {
    const tree = renderer.create(<HOC />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('when {title}', () => {
    const tree = renderer.create(<HOC title={TITLE} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('when {image}', () => {
    const tree = renderer.create(<HOC image={IMAGE} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('addProduct: adding quantity for existing product', () => {
    const index = 1;
    const title = 'Lorem';
    const image = 'lorem.png';
    const price = 23;

    addProduct({ products, total, onData }, index, title, image, price);

    const { total: totalContext, products: productsContext } = providerData.instance().state;
    products[0].quantity = 2;
    expect(totalContext).toEqual(1023);
    expect(productsContext).toEqual(products);
  });

  it('addProduct: adding quantity for other product', () => {
    const index = 3;
    const title = 'Lorem';
    const image = 'lorem.png';
    const price = 45;

    addProduct({ products, total, onData }, index, title, image, price);

    const { total: totalContext, products: productsContext } = providerData.instance().state;
    expect(totalContext).toEqual(1045);
    expect(productsContext).toEqual([...products, { index: 3, title, image, price, quantity: 1 }]); // eslint-disable-line
  });
});
