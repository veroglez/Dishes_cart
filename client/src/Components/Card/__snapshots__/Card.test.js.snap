// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card> renders 1`] = `
<div
  className="card"
>
  <img
    className="image"
  />
  <div
    className="info"
  >
    <p
      className="title"
    />
    <p
      className="price"
    >
      NaN €
    </p>
    <button
      onClick={[Function]}
      type="button"
    >
      Añadir
    </button>
  </div>
</div>
`;

exports[`<Card> when {image} 1`] = `
<div
  className="card"
>
  <img
    className="image"
    src="/image.png"
  />
  <div
    className="info"
  >
    <p
      className="title"
    />
    <p
      className="price"
    >
      NaN €
    </p>
    <button
      onClick={[Function]}
      type="button"
    >
      Añadir
    </button>
  </div>
</div>
`;

exports[`<Card> when {price} 1`] = `
<div
  className="card"
>
  <img
    className="image"
  />
  <div
    className="info"
  >
    <p
      className="title"
    />
    <p
      className="price"
    >
      0.23 €
    </p>
    <button
      onClick={[Function]}
      type="button"
    >
      Añadir
    </button>
  </div>
</div>
`;

exports[`<Card> when {title} 1`] = `
<div
  className="card"
>
  <img
    alt="Lorem ipsum"
    className="image"
  />
  <div
    className="info"
  >
    <p
      className="title"
    >
      Lorem ipsum
    </p>
    <p
      className="price"
    >
      NaN €
    </p>
    <button
      onClick={[Function]}
      type="button"
    >
      Añadir
    </button>
  </div>
</div>
`;
